File: epi_judge_cpp/string_integer_interconversion.cc
#include <iostream>
#include <string>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
using std::string;

string IntToString(int x) {
  long xx = x;
  auto negative = (xx < 0);
  std::deque<char> num;
  xx = abs(xx);
  if (xx == 0) {
    num.push_back('0');
  }
  while (xx) {
    char digit = '0' + xx % 10;
    xx /= 10;
    num.push_front(digit);
  }
  if (negative) {
    num.push_front('-');
  }
  return string(num.begin(), num.end());

  // // bool isneg = false;
  // // if (x < 0) {
  // //   isneg = true;
  // //   x = -x;
  // // }
  // bool isneg = x < 0 ? true : false;
  // unsigned int xx = isneg ? -x : x;
  // string res;
  // do {
  //   res = static_cast<char>(xx % 10 + '0') + res;
  //   xx /= 10;
  // } while (xx > 0);
  // if (isneg) {
  //   res = '-' + res;
  // }
  // return res;
}

int StringToInt(const string &s) {
  long res = 0;
  for (auto &ch : s) {
    if (ch != '-' && ch != '+') {
      res *= 10;
      res += ch - '0';
    }
  }
  res = s.front() == '-' ? -res : res;
  return res;
}
void Wrapper(int x, const string &s) {
  if (stoi(IntToString(x)) != x) {
    throw TestFailure("Int to string conversion failed");
  }

  if (StringToInt(s) != x) {
    throw TestFailure("String to int conversion failed");
  }
}

int main(int argc, char *argv[]) {
  // string foo = IntToString(-2147483648);
  // std::cout << foo  << std::endl;
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"x", "s"};
  return GenericTestMain(args, "string_integer_interconversion.cc",
                         "string_integer_interconversion.tsv", &Wrapper,
                         DefaultComparator{}, param_names);
}

File: epi_judge_python/string_integer_interconversion.py
from test_framework import generic_test
from test_framework.test_failure import TestFailure


def int_to_string(x: int) -> str:
    digits = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
    if x < 0:
        negative = True
        x = -x
    else:
        negative = False
    res = ""
    if x == 0:
        res = "0"
    else:
        while x > 0:
            res = digits[x % 10] + res
            x //= 10
    if negative:
        res = '-' + res
    return res


def string_to_int(s: str) -> int:
    digits = {
        "0": 0,
        "1": 1,
        "2": 2,
        "3": 3,
        "4": 4,
        "5": 5,
        "6": 6,
        "7": 7,
        "8": 8,
        "9": 9,
    }
    base = 1
    negative = False
    if s[0] == "-":
        s = s[1:]
        negative = True
    elif s[0] == "+":
        s = s[1:]
    res = 0
    while s:
        res += digits[s[-1]] * base
        base *= 10
        s = s[:-1]
    if negative:
        res = -res
    return res


def wrapper(x, s):
    if int(int_to_string(x)) != x:
        raise TestFailure("Int to string conversion failed")
    if string_to_int(s) != x:
        raise TestFailure("String to int conversion failed")


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "string_integer_interconversion.py",
            "string_integer_interconversion.tsv",
            wrapper,
        )
    )
